server.go

```diff
@@ -213,31 +265,44 @@ func (s *Server) handleService(mainconn net.Conn, serverName, peeked string, isT
 	log.Println("handling connection from ", mainconn.RemoteAddr())
 
 	conn := GetConn(mainconn, peeked)
-	if err := s.forward(conn, service.Addr, remotePort); err != nil {
-		log.Printf("failed to forward traffic: %v\n", err)
+	if service.ClientSecret != "" {
+		// forward to an active connection
+		if err := s.forwardConnection(conn, s.activeConnections[service.ClientSecret]); err != nil {
```

Comment:
same remark as before, delay the error handling after the if/else block so you only do it once.
Here you don't print the error for the line 271 for example.